### DISCUSSION TITLE
TFSwitch --latest should have an equivalent that will conform to version constraints in tf files

### DISCUSSION THREAD

**Discussion by itsalix** [warrensbox/terraform-switcher#474] [Ideas]
I am in a situation where I am automatically running tfswitch for users based on different scenarios. 
They can either define a constraint in their terraform block and have tfswitch download the matching version or do nothing and have tfswitch download the latest version available. The issue is, due to the behaviour of running tfswitch --latest, it acknowledges the version constraint in the files but then ignores it to download the latest existing version, as shown below.
`
❯ tfswitch --latest
11:42:33.966 INFO Reading version from Terraform module at "."  
11:42:34.063 INFO Reading required version from constraint: "~> 1.5.0"  
11:42:34.064 INFO Matched version: "1.5.7"  
11:42:34.158 INFO Switched terraform to version "1.8.5"
`

I don't want to read in their terraform configuration BEFORE calling tfswitch because tfswitch already has that functionality built in, however I need to know that information before I run the tfswitch command, as it decides whether I run `tfswitch` or `tfswitch --latest` 
It would be good if there was a flag to not have the interactive selection when not passing a version and not having a version constraint in your tf code, i.e. a `--latest` flag that does not ignore the constraints in the code, but instead defaults to the latest if no constraint is found. `--latest-within-constraint` if that makes sense.

I may be missing something obvious that meets my use case 😅 but from reading through docs and playing around at the CLI, I haven't seen anything yet

**Comment by yermulnik** [comment]
Would the below work for your use case?
```shell
tfswitch -d $(tfswitch -U -g ERROR | awk 'END {print $NF}' | sed 's/\x1b\[[0-9;]*[mGKHF]//g')
```
* `tfswitch -U -g ERROR` — output the latest stable version (and decrease logging)
* `awk 'END {print $NF}'` — print only last bit of last line of input
* `sed 's/\x1b\[[0-9;]*[mGKHF]//g'` — strip ascii-escape sequences (this should be not needed when we get to implement #462)
* `tfswitch -d` — use default version if no other versions could be detected

**Reply by itsalix** [reply]
Possibly, I am wrapping tfswitch with a Go app, so I would like to avoid relying too heavily on bash, but this seems achievable running tfswitch twice which I hadn't really considered. Thanks 😄  

I do still think it could be a good native feature however.

**Reply by yermulnik** [reply]
> I am wrapping tfswitch with a Go app

Wrapping as an external process or utilizing as Go library? For the latter case we're aiming to improve code and provide `tfswitch` as an external lib in a more reliable way rather than we've got now. Cc @MatthewJohn

> I do still think it could be a good native feature however.

Please create an issue for this feature request and we will work on it as time and human power permit. Thanks.

**Reply by itsalix** [reply]
Wrapping as an external process for now, always interested to see the improvements for utilising it as a library though.

> Please create an issue for this feature request and we will work on it as time and human power permit. Thanks.

Will do 🫡

**Reply by yermulnik** [reply]
Thank you
Linking it for tracking: https://github.com/warrensbox/terraform-switcher/issues/475

**Reply by yermulnik** [reply]
WIP: https://github.com/warrensbox/terraform-switcher/pull/536

**Reply by yermulnik** [reply]
I'm closing this discussion as it's being worked on at the moment.
Please let me know if I'd better of re-open it for whatever reason.

**Reply by yermulnik** [reply]
JFYI: #631
